File: mvvmbase/src/main/java/com/bing/mvvmbase/base/viewpager/BaseBottomNavigationActivity.java
package com.bing.mvvmbase.base.viewpager;

import android.os.Bundle;
import android.view.MenuItem;

import com.bing.mvvmbase.base.BaseViewModel;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.lifecycle.AndroidViewModel;
import androidx.viewpager.widget.ViewPager;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BaseBottomNavigationActivity<DB extends ViewDataBinding, VM extends BaseViewModel> extends AppCompatActivity {
	protected DB mBinding;
	protected VM mViewModel;
	protected CompositeDisposable mCompositeDisposable = new CompositeDisposable();
	protected ViewPager mViewPager;
	protected FragmentPagerAdapter mFragmentPagerAdapter;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		onCreateFirst();
		initViewModel();
		getLifecycle().addObserver(mViewModel);
		mBinding = DataBindingUtil.setContentView(this, layoutId());
		initViewPagerAndBottomNav();
		bindAndObserve();
	}

	protected void onCreateFirst() {

	}

	protected abstract void initViewModel();
	protected abstract int layoutId();

	protected void initViewPagerAndBottomNav() {
		mViewPager = viewpager();
		mFragmentPagerAdapter = new BaseFragmentPagerAdapter(getSupportFragmentManager(), initFragments());
		mViewPager.setAdapter(mFragmentPagerAdapter);
		navigationView().setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);
	}

	protected abstract ViewPager viewpager();
	protected abstract BottomNavigationView navigationView();
	protected abstract List<Fragment> initFragments();
	protected abstract void bindAndObserve();

	protected void addDisposable(Disposable disposable) {
		mCompositeDisposable.add(disposable);
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		mCompositeDisposable.dispose();
	}


	private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
			= new BottomNavigationView.OnNavigationItemSelectedListener() {

		@Override
		public boolean onNavigationItemSelected(@NonNull MenuItem item) {
			return navigationItemSelected(item);
		}
	};

	protected abstract boolean navigationItemSelected(@NonNull MenuItem item);
}

File: mvvmbase/src/main/java/com/bing/mvvmbase/base/viewpager/BaseBottomNavigationActivity.kt
package com.bing.mvvmbase.base.viewpager

import android.os.Bundle
import android.view.MenuItem
import android.view.View

import com.bing.mvvmbase.base.BaseViewModel
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BaseBottomNavigationActivity<DB : ViewDataBinding, VM : BaseViewModel> : AppCompatActivity(), View.OnClickListener {
        lateinit var mBinding:DB
        lateinit var mViewModel: VM
        val mCompositeDisposable = CompositeDisposable()
        lateinit var mViewPager: ViewPager
        lateinit var mFragmentPagerAdapter: FragmentPagerAdapter

        private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item -> navigationItemSelected(item) }

        override fun onCreate(savedInstanceState: Bundle?) {
                super.onCreate(savedInstanceState)
                onCreateFirst()
                mBinding = DataBindingUtil.setContentView(this, layoutId())
                initViewModel()
                lifecycle.addObserver(mViewModel)
                initViewPagerAndBottomNav()
                bindAndObserve()
        }

        open fun onCreateFirst() {

        }

        protected abstract fun layoutId(): Int
        abstract fun initViewModel()

        open fun initViewPagerAndBottomNav() {
                mViewPager = viewpager()
                mFragmentPagerAdapter = BaseFragmentPagerAdapter(supportFragmentManager, initFragments())
                mViewPager.adapter = mFragmentPagerAdapter
                navigationView().setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)
        }

        protected abstract fun viewpager(): ViewPager
        protected abstract fun navigationView(): BottomNavigationView
        protected abstract fun initFragments(): List<Fragment>
        protected abstract fun bindAndObserve()

        protected fun addOnClickListener(vararg views: View) {
                for (view in views) {
                        view.setOnClickListener(this)
                }
        }

        override fun onClick(v: View) {

        }

        protected fun addDisposable(disposable: Disposable) {
                mCompositeDisposable.add(disposable)
        }

        override fun onDestroy() {
                mCompositeDisposable.dispose()
                super.onDestroy()
        }

        protected abstract fun navigationItemSelected(item: MenuItem): Boolean
}
